tsdb/exemplar.go

```diff
@@ -133,62 +163,66 @@ Outer:
 	return false
 }
 
-// indexGc takes the circularBufferEntry that will be overwritten and updates the
-// storages index for that entries labelset if necessary.
-func (ce *CircularExemplarStorage) indexGc(cbe *circularBufferEntry) {
-	if cbe == nil {
-		return
-	}
-
-	l := cbe.seriesLabels.String()
-	i := cbe.next
-	if i == -1 {
-		delete(ce.index, l)
-		return
-	}
-
-	ce.index[l] = &indexEntry{i, ce.index[l].last}
-}
-
 func (ce *CircularExemplarStorage) AddExemplar(l labels.Labels, e exemplar.Exemplar) error {
 	seriesLabels := l.String()
+
+	// TODO(bwplotka): This lock can lock all scrapers, there might high contention on this on scale.
+	// Optimize by moving the lock to be per series (& benchmark it).
 	ce.lock.Lock()
 	defer ce.lock.Unlock()
 
 	idx, ok := ce.index[seriesLabels]
 	if !ok {
-		ce.indexGc(ce.exemplars[ce.nextIndex])
-		// Default the next value to -1 (which we use to detect that we've iterated through all exemplars for a series in Select)
-		// since this is the first exemplar stored for this series.
-		ce.exemplars[ce.nextIndex] = &circularBufferEntry{
-			exemplar:     e,
-			seriesLabels: l,
-			next:         -1}
-		ce.index[seriesLabels] = &indexEntry{ce.nextIndex, ce.nextIndex}
-		ce.nextIndex = (ce.nextIndex + 1) % len(ce.exemplars)
-		return nil
-	}
+		ce.index[seriesLabels] = &indexEntry{first: ce.nextIndex}
+	} else {
+		// Check for duplicate vs last stored exemplar for this series.
+		// NB these are expected, add appending them is a no-op.
+		if ce.exemplars[idx.last].exemplar.Equals(e) {
+			return nil
+		}
 
-	// Check for duplicate vs last stored exemplar for this series.
-	// NB these are expected, add appending them is a no-op.
-	if ce.exemplars[idx.last].exemplar.Equals(e) {
-		return nil
-	}
+		if e.Ts <= ce.exemplars[idx.last].exemplar.Ts {
+			ce.outOfOrderExemplars.Inc()
+			return storage.ErrOutOfOrderExemplar
+		}
 
-	if e.Ts <= ce.exemplars[idx.last].exemplar.Ts {
-		ce.outOfOrderExemplars.Inc()
-		return storage.ErrOutOfOrderExemplar
+		ce.exemplars[ce.index[seriesLabels].last].next = ce.nextIndex
 	}
-	ce.indexGc(ce.exemplars[ce.nextIndex])
-	ce.exemplars[ce.nextIndex] = &circularBufferEntry{
-		exemplar:     e,
-		seriesLabels: l,
-		next:         -1,
+
+	if prev := ce.exemplars[ce.nextIndex]; prev == nil {
+		ce.exemplars[ce.nextIndex] = &circularBufferEntry{}
+	} else {
+		// There exists exemplar already on this ce.nextIndex entry, drop it, to make place
+		// for others.
+		prevLabels := prev.seriesLabels.String()
+		if prev.next == -1 {
+			// Last item for this series, remove index entry.
+			delete(ce.index, prevLabels)
+		} else {
+			ce.index[prevLabels].first = prev.next
+		}
 	}
 
-	ce.exemplars[ce.index[seriesLabels].last].next = ce.nextIndex
+	// Default the next value to -1 (which we use to detect that we've iterated through all exemplars for a series in Select)
+	// since this is the first exemplar stored for this series.
+	ce.exemplars[ce.nextIndex].exemplar = e
```

Comment:
nice, this probably gets rid of a lot of allocs

Reply:
👍🏽